relays/lib-substrate-relay/src/messages_source.rs

```diff
@@ -360,10 +361,93 @@ where
 			.await
 	}
 
-	async fn require_target_header_on_source(&self, id: TargetHeaderIdOf<MessageLaneAdapter<P>>) {
+	async fn require_target_header_on_source(
+		&self,
+		id: TargetHeaderIdOf<MessageLaneAdapter<P>>,
+	) -> Result<Option<Self::BatchTransaction>, SubstrateError> {
 		if let Some(ref target_to_source_headers_relay) = self.target_to_source_headers_relay {
+			if P::SourceBatchCallBuilder::BATCH_CALL_SUPPORTED {
+				return BatchConfirmationTransaction::<P>::new(self.clone(), id).await.map(Some)
+			}
+
 			target_to_source_headers_relay.require_more_headers(id.0).await;
 		}
+
+		Ok(None)
+	}
+}
+
+/// Batch transaction that brings target headers + and delivery confirmations to the source node.
+pub struct BatchConfirmationTransaction<P: SubstrateMessageLane> {
+	messages_source: SubstrateMessagesSource<P>,
+	proved_header: TargetHeaderIdOf<MessageLaneAdapter<P>>,
+	prove_calls: Vec<CallOf<P::SourceChain>>,
+}
+
+impl<P: SubstrateMessageLane> BatchConfirmationTransaction<P> {
+	async fn new(
+		messages_source: SubstrateMessagesSource<P>,
+		required_target_header_on_source: TargetHeaderIdOf<MessageLaneAdapter<P>>,
+	) -> Result<Self, SubstrateError> {
+		let (proved_header, prove_calls) = messages_source
+			.target_to_source_headers_relay
+			.as_ref()
+			.expect("BatchConfirmationTransaction is only created when target_to_source_headers_relay is Some; qed")
```

Comment:
I'd totally agree with you, if this function (`fn new`) would have been public. But it is inaccessible outside of the module (i.e. not `pub`) and the proof describes the context where it is created. I could move the `new` code to the place where it is created (a couple of lines above) => remove this expect, but since the creation code isn't trivial anymore, I've decided to extract it into a separate method. Feel free to insist - then I'll move it back :)

Reply:
I think I would move the code to the place where the struct is created than. But it's not a strong preference. However you prefer.